File: SRC/dlanv2.f
      SUBROUTINE DLANV2( A, B, C, D, RT1R, RT1I, RT2R, RT2I, CS, SN )

*  -- LAPACK auxiliary routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      double             A, B, C, CS, D, RT1I, RT1R, RT2I, RT2R, SN;
      // ..

*  =====================================================================

      // .. Parameters ..
      double             ZERO, HALF, ONE, TWO;
      PARAMETER          ( ZERO = 0.0D+0, HALF = 0.5D+0, ONE = 1.0D+0, TWO = 2.0D0 )
      double             MULTPL;
      PARAMETER          ( MULTPL = 4.0D+0 )
      // ..
      // .. Local Scalars ..
      double             AA, BB, BCMAX, BCMIS, CC, CS1, DD, EPS, P, SAB, SAC, SCALE, SIGMA, SN1, TAU, TEMP, Z, SAFMIN, SAFMN2, SAFMX2;
      int                COUNT;
      // ..
      // .. External Functions ..
      double             DLAMCH, DLAPY2;
      // EXTERNAL DLAMCH, DLAPY2
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC ABS, MAX, MIN, SIGN, SQRT
      // ..
      // .. Executable Statements ..

      SAFMIN = DLAMCH( 'S' )
      EPS = DLAMCH( 'P' )
      SAFMN2 = DLAMCH( 'B' )**INT( LOG( SAFMIN / EPS ) / LOG( DLAMCH( 'B' ) ) / TWO )
      SAFMX2 = ONE / SAFMN2
      IF( C.EQ.ZERO ) THEN
         CS = ONE
         SN = ZERO

      ELSE IF( B.EQ.ZERO ) THEN

         // Swap rows and columns

         CS = ZERO
         SN = ONE
         TEMP = D
         D = A
         A = TEMP
         B = -C
         C = ZERO

      ELSE IF( ( A-D ).EQ.ZERO .AND. SIGN( ONE, B ).NE.SIGN( ONE, C ) ) THEN
         CS = ONE
         SN = ZERO

      ELSE

         TEMP = A - D
         P = HALF*TEMP
         BCMAX = MAX( ABS( B ), ABS( C ) )
         BCMIS = MIN( ABS( B ), ABS( C ) )*SIGN( ONE, B )*SIGN( ONE, C )
         SCALE = MAX( ABS( P ), BCMAX )
         Z = ( P / SCALE )*P + ( BCMAX / SCALE )*BCMIS

         // If Z is of the order of the machine accuracy, postpone the
         // decision on the nature of eigenvalues

         IF( Z.GE.MULTPL*EPS ) THEN

            // Real eigenvalues. Compute A and D.

            Z = P + SIGN( SQRT( SCALE )*SQRT( Z ), P )
            A = D + Z
            D = D - ( BCMAX / Z )*BCMIS

            // Compute B and the rotation matrix

            TAU = DLAPY2( C, Z )
            CS = Z / TAU
            SN = C / TAU
            B = B - C
            C = ZERO

         ELSE

            // Complex eigenvalues, or real (almost) equal eigenvalues.
            // Make diagonal elements equal.

            COUNT = 0
            SIGMA = B + C
   10       CONTINUE
            COUNT = COUNT + 1
            SCALE = MAX( ABS(TEMP), ABS(SIGMA) )
            IF( SCALE.GE.SAFMX2 ) THEN
               SIGMA = SIGMA * SAFMN2
               TEMP = TEMP * SAFMN2
               IF (COUNT .LE. 20) GOTO 10
            END IF
            IF( SCALE.LE.SAFMN2 ) THEN
               SIGMA = SIGMA * SAFMX2
               TEMP = TEMP * SAFMX2
               IF (COUNT .LE. 20) GOTO 10
            END IF
            P = HALF*TEMP
            TAU = DLAPY2( SIGMA, TEMP )
            CS = SQRT( HALF*( ONE+ABS( SIGMA ) / TAU ) )
            SN = -( P / ( TAU*CS ) )*SIGN( ONE, SIGMA )

            // Compute [ AA  BB ] = [ A  B ] [ CS -SN ]
                    // [ CC  DD ]   [ C  D ] [ SN  CS ]

            AA = A*CS + B*SN
            BB = -A*SN + B*CS
            CC = C*CS + D*SN
            DD = -C*SN + D*CS

            // Compute [ A  B ] = [ CS  SN ] [ AA  BB ]
                    // [ C  D ]   [-SN  CS ] [ CC  DD ]

            A = AA*CS + CC*SN
            B = BB*CS + DD*SN
            C = -AA*SN + CC*CS
            D = -BB*SN + DD*CS

            TEMP = HALF*( A+D )
            A = TEMP
            D = TEMP

            IF( C.NE.ZERO ) THEN
               IF( B.NE.ZERO ) THEN
                  IF( SIGN( ONE, B ).EQ.SIGN( ONE, C ) ) THEN

                     // Real eigenvalues: reduce to upper triangular form

                     SAB = SQRT( ABS( B ) )
                     SAC = SQRT( ABS( C ) )
                     P = SIGN( SAB*SAC, C )
                     TAU = ONE / SQRT( ABS( B+C ) )
                     A = TEMP + P
                     D = TEMP - P
                     B = B - C
                     C = ZERO
                     CS1 = SAB*TAU
                     SN1 = SAC*TAU
                     TEMP = CS*CS1 - SN*SN1
                     SN = CS*SN1 + SN*CS1
                     CS = TEMP
                  END IF
               ELSE
                  B = -C
                  C = ZERO
                  TEMP = CS
                  CS = -SN
                  SN = TEMP
               END IF
            END IF
         END IF

      END IF

      // Store eigenvalues in (RT1R,RT1I) and (RT2R,RT2I).

      RT1R = A
      RT2R = D
      IF( C.EQ.ZERO ) THEN
         RT1I = ZERO
         RT2I = ZERO
      ELSE
         RT1I = SQRT( ABS( B ) )*SQRT( ABS( C ) )
         RT2I = -RT1I
      END IF
      RETURN

      // End of DLANV2

      END

File: lib/src/dlanv2.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/dlapy2.dart';
import 'package:dart_lapack/src/intrinsics/sign.dart';
import 'package:dart_lapack/src/install/dlamch.dart';

void dlanv2(
  final Box<double> A,
  final Box<double> B,
  final Box<double> C,
  final Box<double> D,
  final Box<double> RT1R,
  final Box<double> RT1I,
  final Box<double> RT2R,
  final Box<double> RT2I,
  final Box<double> CS,
  final Box<double> SN,
) {
// -- LAPACK auxiliary routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  const ZERO = 0.0, HALF = 0.5, ONE = 1.0, TWO = 2.0;
  const MULTPL = 4.0;
  double AA,
      BB,
      BCMAX,
      BCMIS,
      CC,
      CS1,
      DD,
      EPS,
      P,
      SAB,
      SAC,
      SCALE,
      SIGMA = 0,
      SN1,
      TAU,
      TEMP,
      Z,
      SAFMIN,
      SAFMN2,
      SAFMX2;
  int COUNT = 0;

  SAFMIN = dlamch('S');
  EPS = dlamch('P');
  SAFMN2 = pow(
    dlamch('B'),
    (log(SAFMIN / EPS) / log(dlamch('B')) ~/ TWO),
  ).toDouble();
  SAFMX2 = ONE / SAFMN2;
  if (C.value == ZERO) {
    CS.value = ONE;
    SN.value = ZERO;
  } else if (B.value == ZERO) {
    // Swap rows and columns

    CS.value = ZERO;
    SN.value = ONE;
    TEMP = D.value;
    D.value = A.value;
    A.value = TEMP;
    B.value = -C.value;
    C.value = ZERO;
  } else if ((A.value - D.value) == ZERO &&
      sign(ONE, B.value) != sign(ONE, C.value)) {
    CS.value = ONE;
    SN.value = ZERO;
  } else {
    TEMP = A.value - D.value;
    P = HALF * TEMP;
    BCMAX = max(B.value.abs(), C.value.abs());
    BCMIS = min(B.value.abs(), C.value.abs()) *
        sign(ONE, B.value) *
        sign(ONE, C.value);
    SCALE = max(P.abs(), BCMAX);
    Z = (P / SCALE) * P + (BCMAX / SCALE) * BCMIS;

    // If Z is of the order of the machine accuracy, postpone the
    // decision on the nature of eigenvalues

    if (Z >= MULTPL * EPS) {
      // Real eigenvalues. Compute A and D.

      Z = P + sign(sqrt(SCALE) * sqrt(Z), P);
      A.value = D.value + Z;
      D.value -= (BCMAX / Z) * BCMIS;

      // Compute B and the rotation matrix

      TAU = dlapy2(C.value, Z);
      CS.value = Z / TAU;
      SN.value = C.value / TAU;
      B.value -= C.value;
      C.value = ZERO;
    } else {
      // Complex eigenvalues, or real (almost) equal eigenvalues.
      // Make diagonal elements equal.

      COUNT = 0;
      SIGMA = B.value + C.value;
      while (true) {
        COUNT++;
        SCALE = max(TEMP.abs(), SIGMA.abs());
        if (SCALE >= SAFMX2) {
          SIGMA *= SAFMN2;
          TEMP *= SAFMN2;
          if (COUNT <= 20) continue;
        }
        if (SCALE <= SAFMN2) {
          SIGMA *= SAFMX2;
          TEMP *= SAFMX2;
          if (COUNT <= 20) continue;
        }
        break;
      }
      P = HALF * TEMP;
      TAU = dlapy2(SIGMA, TEMP);
      CS.value = sqrt(HALF * (ONE + SIGMA.abs() / TAU));
      SN.value = -(P / (TAU * CS.value)) * sign(ONE, SIGMA);

      // Compute [ AA  BB ] = [ A  B ] [ CS -SN ]
      // [ CC  DD ]   [ C  D ] [ SN  CS ]

      AA = A.value * CS.value + B.value * SN.value;
      BB = -A.value * SN.value + B.value * CS.value;
      CC = C.value * CS.value + D.value * SN.value;
      DD = -C.value * SN.value + D.value * CS.value;

      // Compute [ A  B ] = [ CS  SN ] [ AA  BB ]
      // [ C  D ]   [-SN  CS ] [ CC  DD ]

      A.value = AA * CS.value + CC * SN.value;
      B.value = BB * CS.value + DD * SN.value;
      C.value = -AA * SN.value + CC * CS.value;
      D.value = -BB * SN.value + DD * CS.value;

      TEMP = HALF * (A.value + D.value);
      A.value = TEMP;
      D.value = TEMP;

      if (C.value != ZERO) {
        if (B.value != ZERO) {
          if (sign(ONE, B.value) == sign(ONE, C.value)) {
            // Real eigenvalues: reduce to upper triangular form

            SAB = sqrt(B.value.abs());
            SAC = sqrt(C.value.abs());
            P = sign(SAB * SAC, C.value);
            TAU = ONE / sqrt((B.value + C.value).abs());
            A.value = TEMP + P;
            D.value = TEMP - P;
            B.value -= C.value;
            C.value = ZERO;
            CS1 = SAB * TAU;
            SN1 = SAC * TAU;
            TEMP = CS.value * CS1 - SN.value * SN1;
            SN.value = CS.value * SN1 + SN.value * CS1;
            CS.value = TEMP;
          }
        } else {
          B.value = -C.value;
          C.value = ZERO;
          TEMP = CS.value;
          CS.value = -SN.value;
          SN.value = TEMP;
        }
      }
    }
  }

  // Store eigenvalues in (RT1R,RT1I) and (RT2R,RT2I).

  RT1R.value = A.value;
  RT2R.value = D.value;
  if (C.value == ZERO) {
    RT1I.value = ZERO;
    RT2I.value = ZERO;
  } else {
    RT1I.value = sqrt(B.value.abs()) * sqrt(C.value.abs());
    RT2I.value = -RT1I.value;
  }
}
